### PR TITLE
Stabilize const_fn_fn_ptr_basics, const_fn_trait_bound, and const_impl_trait

### PR DESCRIPTION
**Pull request by eholk**
# Stabilization Report

This PR serves as a request for stabilization for three const evaluation features:

1. `const_fn_fn_ptr_basics`
2. `const_fn_trait_bound`
3. `const_impl_trait`

These are being stabilized together because they are relatively minor and related updates to existing functionality.

## `const_fn_fn_ptr_basics`

Allows creating, passing, and casting function pointers in a `const fn`.

The following is an example of what is now allowed:

```rust
const fn get_function() -> fn() {
    fn foo() {
        println!("Hello, World!");
    }
    
    foo
}
```

Casts between function pointer types are allowed, as well as transmuting from integers:

```rust
const fn get_function() -> fn() {
    unsafe {
        std::mem::transmute(0x1234usize)
    }
}
```

However, casting from a function pointer to an integer is not allowed:

```rust 
const fn fn_to_usize(f: fn()) -> usize {
    f as usize  //~ pointers cannot be cast to integers during const eval
}
```

Calling function pointers is also not allowed.

```rust
const fn call_fn_ptr(f: fn()) {
    f() //~ function pointers are not allowed in const fn 
}
```

### Test Coverage

The following tests include code that exercises this feature:

- `src/test/ui/consts/issue-37550.rs`
- `src/test/ui/consts/issue-46553.rs`
- `src/test/ui/consts/issue-56164.rs`
- `src/test/ui/consts/min_const_fn/allow_const_fn_ptr_run_pass.rs`
- `src/test/ui/consts/min_const_fn/cast_fn.rs`
- `src/test/ui/consts/min_const_fn/cmp_fn_pointers.rs`

## `const_fn_trait_bound`

Allows trait bounds in `const fn`. Additionally, this feature allows creating and passing `dyn Trait` objects.

Examples such as the following are allowed by this feature:

```rust
const fn do_thing<T: Foo>(_x: &T) {
    // ...
}
```

Previously only `Sized` was allowed as a trait bound.

There is no way to call methods from the trait because trait methods cannot currently be marked as const. Allowing trait bounds in const functions does allow the const function to use the trait's associated types and constants.

This feature also allowes `dyn Trait` types. These work equivalently to non-const code. Similar to other pointers in const code, the value of a `dyn Trait` pointer cannot be observed.

Note that due to https://github.com/rust-lang/rust/issues/90912, it was already possible to do the example above as follows:

```rust 
const fn do_thing<T>(_x: &T) where (T,): Foo {
    // ...
}
```

### Test Coverage

The following tests include code that exercises `const_fn_trait_bound`:

- `src/test/ui/consts/const-fn.rs`
- `src/test/ui/consts/issue-88071.rs`
- `src/test/ui/consts/min_const_fn/min_const_fn.rs`
- `src/test/ui/consts/min_const_fn/min_const_fn_dyn.rs`
- `src/test/ui/nll/issue-55825-const-fn.rs`
- Many of the tests in `src/test/ui/rfc-2632-const-trait-impl/` also exercise this feature.

## `const_impl_trait`

Allows argument and return position `impl Trait` in a `const fn`, such as in the following example:

```rust 
const fn do_thing(x: impl Foo) -> impl Foo {
    x
}
```

Similar to generic parameters and function pointers, this allows the creation of such opaque types, but not doing anything with them beyond accessing associated types and constants.

### Test Coverage

The following tests exercise this feature:

- `src/test/ui/type-alias-impl-trait/issue-53096.rs`
- `src/test/ui/type-alias-impl-trait/issue-53678-generator-and-const-fn.rs`

## Documentation

These features are documented along with the other const evaluation features in the Rust Reference at https://doc.rust-lang.org/stable/reference/const_eval.html.

There is a PR that updates this documentation to reflect the capabilities enabled by these features at https://github.com/rust-lang/reference/pull/1166.

Tracking issues: #57563, #63997, #93706

### CODE REVIEW

**Comment by CryZe** [comment]
Maybe weird question, but the RFC for this was closed, not accepted, how can this be stabilized?

**Comment by fee1-dead** [comment]
Oh, you beat me to it! Nice work!

**Comment by eholk** [comment]
I have a documentation PR at https://github.com/rust-lang/reference/pull/1166.

**Comment by fee1-dead** [comment]
Please also remove `ops::ty::DynTrait` as well as `ops::ty::TraitBoundNotConst`. `dyn` should be part of the stabilization and the latter was obsolete after I introduced `~const`.

**Comment by RalfJung** [comment]
To be clear, what is being "stabilized" here is simply mentioning the types `fn()` and `dyn Trait` in `const fn` -- there is no way to actually do anything with these types during CTFE, other than pass them around and use them to initialized a `static`/`const`.

Actually having function pointers or trait objects that can be called during CTFE is future work, and will require new syntax.

File: compiler/rustc_const_eval/src/transform/check_consts/check.rs [hunk 1]
```diff
@@ -407,6 +390,7 @@ impl<'mir, 'tcx> Checker<'mir, 'tcx> {
                     | ty::PredicateKind::ConstEvaluatable(..)
                     | ty::PredicateKind::ConstEquate(..)
                     | ty::PredicateKind::OpaqueType(..)
+                    | ty::PredicateKind::Trait(..)
```
Comment (lcnr) [inline compiler/rustc_const_eval/src/transform/check_consts/check.rs:393]: this whole function is now a noop, isn't it?

File: compiler/rustc_const_eval/src/transform/check_consts/check.rs [hunk 2]
```diff
@@ -625,7 +588,9 @@ impl<'tcx> Visitor<'tcx> for Checker<'_, 'tcx> {
                 ),
                 _,
                 _,
-            ) => self.check_op(ops::FnPtrCast),
```
[the remark this answers is not in the document]
Comment (eholk) [reply compiler/rustc_const_eval/src/transform/check_consts/check.rs:628]: Correct, this change allows function pointers to be passed around but not called. I updated the PR description to clarify this.

File: compiler/rustc_const_eval/src/transform/check_consts/check.rs (hunk 1 above)
Comment (eholk) [reply compiler/rustc_const_eval/src/transform/check_consts/check.rs]: Good catch. I removed it.

**Comment by eholk** [comment]
> Please also remove `ops::ty::DynTrait` as well as `ops::ty::TraitBoundNotConst`. `dyn` should be part of the stabilization and the latter was obsolete after I introduced `~const`.

Done! `DynTrait` was already gone, but I just got rid of `TraitBoundNotConst` too.

**Comment by oli-obk** [comment]
I think my comment on zulip got drowned out by the other thread XD. Repeating it here:

> Awesome! Looked over it and it lgtm. If you want you can also throw const_impl_trait out in the same PR.

> Wrt documentation, all that needs updating is the reference in  https://doc.rust-lang.org/stable/reference/const_eval.html?highlight=Const#const-functions

**Comment by oli-obk** [comment]
> To be clear, what is being "stabilized" here is simply mentioning the types `fn()` and `dyn Trait` in `const fn` -- there is no way to actually do anything with these types during CTFE, other than pass them around and use them to initialize a `static`/`const`.
> 
> Actually having function pointers or trait objects that can be called during CTFE is future work, and will require new syntax.

Repeating this comment, as I'm now nominating this PR for @rust-lang/lang to consider. In the [const eval meeting](https://github.com/rust-lang/lang-team/blob/master/design-meeting-minutes/2021-12-22-Const-eval-update.md) we had right before the new year and in the [zulip discussion](https://zulip-archive.rust-lang.org/stream/213817-t-lang/topic/const-if-const.20syntax.20(.60~const.60).html#265842213) afterwards, there seemed resounding support for this PR's change.

The PR fixes https://github.com/rust-lang/rust/issues/90912

**Comment by Mark-Simulacrum** [comment]
The PR title suggests this is aiming to stabilize both const_fn_trait_bound and const_fn_fn_ptr_basics, the latter of which seems to be what your comment discusses, whereas const_fn_trait_bound is not addressed? Are these features coupled in some way, or is it just being stabilized in the same PR for convenience?

Regardless, I think it would be good to write a more complete description of the changes here into the PR description -- just because this is stabilizing 'nicer syntax' rather than truly new functionality IMO does not mean that we should side step a stabilization report for this new feature, even if it is relatively simple -- some questions that I have:

* What are our semantics for creating fn-ptr or dyn Trait values in const fn? Is behavior here equivalent to behavior in regular functions?
* Can the value of the fn-ptr or dyn Trait be observed? (e.g., by transmuting)
* Can you cast from usize or similar to an fn-ptr (or in theory dyn Trait) in const fn?

IOW, it seems like it would be a reasonable step to stabilize these as-if we were in a world where they were entirely inaccessible in const fn, since that is the blessed world from a language perspective.

For const fn trait bounds, I think a similar approach of 'from scratch' stabilization, rather than a delta with the already stable language would be good. Obviously, some (all?) of the decisions have already been fairly fixed in place, but having that stabilization report would let us discuss whether some of them are in hindsight a bad idea, and potentially propose either edition changes or otherwise break the behavior in the future; skipping the stabilization report doesn't really do us much good.

---

More on a T-compiler perspective -- it sounds like the stabilization here (via the workarounds) was done largely by accident -- have we expanded test coverage or considered what can be done to avoid future mistakes? Is there some extra validation we should try to do?

**Comment by eholk** [comment]
> Awesome! Looked over it and it lgtm. If you want you can also throw const_impl_trait out in the same PR.

Ah, sorry I missed this part. I just pushed another commit that includes `const_impl_trait`.

**Comment by nikomatsakis** [comment]
I'll second what @Mark-Simulacrum wrote, a stabilization report would be great. I will *also* say that in some recent conversations, @tmandry and I were landing on designs that really wanted the ability to write a `const fn` that returns a function pointer, with the assumption that the resulting `fn` pointer would be something the compiler can statically link to (or at least know that it cannot). So I'm excited to see this being proposed for stabilization and, separately, I'd like to talk over with @oli-obk and make sure that the plans @tmandry and I are dreaming up align with overall reality.

**Comment by eholk** [comment]
I updated the PR description with a stabilization report. Hopefully I addressed most of the important questions. Let me know what questions remain!

**Comment by RalfJung** [comment]
> Casts between function pointer types are allowed, including unsafe casts from integers:

That's a transmute, not a cast, at least as far as the syntax and semantics of Rust are concerned -- so it seems confusing to call this a "cast".

**Comment by eholk** [comment]
> That's a transmute, not a cast, at least as far as the syntax and semantics of Rust are concerned -- so it seems confusing to call this a "cast".

I updated the wording to call that a transmute. I think of transmute as being an unsafe case, at least in effect, even if that's not Rust terminology so that's probably why I worded it that way at first.

**Comment by RalfJung** [comment]
Yeah, it's a lot like reinterpret_cast in C++, but I think in Rust we usually only call the `as` operator "cast".

Thanks for the stabilization report, looks good!

**Comment by CraftSpider** [comment]
Question: Will this stabilization allow the following code?
```
#![feature(const_fn_fn_ptr_basics)]

const fn foo() -> fn(&mut i32) {
    fn bar(_: &mut i32) {}

    bar
}
```
Currently this fails due to requiring the `const_mut_refs` feature, but if there's no way to call the provided function pointers, it seems odd to deny them containing mutable references. Unless there are future plans to allow 'implicit' const function pointers?

**Comment by fee1-dead** [comment]
That is probably because of conservative checking. I am more in favor of stabilizing `const_mut_refs` partially and include this one. https://github.com/rust-lang/rust/issues/57349#issuecomment-1024432438

**Comment by nbdd0121** [comment]
> That is probably because of conservative checking. I am more in favor of stabilizing `const_mut_refs` partially and include this one.

I personally would be considering that a "bugfix" rather than "partial stabilisation", i.e. the fact that `fn(&mut i32)` is considered as using mut refs should be considered a bug. Given
```rust
struct MyMutRef<'a, T>(&'a mut T);

const fn accept_my_mut_ref(x: MyMutRef<'_, usize>) {}
```
is accepted today there are no reason that `fn(&mut i32)` should be rejected when appearing in signature.

**Comment by nikomatsakis** [comment]
@rfcbot fcp merge

I think this seems like a good step forward, and a logical extension to what we've got supported so far. I'd like to kickstart the fcp process.

**Comment by oli-obk** [comment]
@rfcbot cancel the PR had the wrong team's label

**Comment by nikomatsakis** [comment]
@rfcbot fcp cancel

**Comment by eholk** [comment]
bors try

**Review by wesleywiser — APPROVED** [review]

**Comment by wesleywiser** [comment]
The CI failure looks to be #93784. Since that isn't related to this PR, I'm going to go ahead and r+. 

@bors r+

**Comment by jplatte** [comment]
I'm doubtful about the `const_fn_trait_bound` stabilization going in the right direction, #93706 is probably the right place to discuss.

**Pull request merged**